src/box/call_rt_user_privilege.c

```diff
@@ -0,0 +1,167 @@
+#include "call_user_privilege.h"
+#include "diag.h"
+#include "assoc.h"
+#include "user.h"
+#include "session.h"
+#include "../lua/init.h" /* tarantool_lua_is_builtin_global */
+
+static struct mh_priv_t *user_privilege_cache = NULL;
+static struct mh_i32ptr_t *user_runtime_universe = NULL;
+
+void
+user_privilege_cache_init(void)
+{
+	user_privilege_cache = mh_priv_new();
+	user_runtime_universe = mh_i32ptr_new();
+}
+
+void
+user_privilege_cache_destroy(void)
+{
+	mh_priv_delete(user_privilege_cache);
+	mh_i32ptr_delete(user_runtime_universe);
+}
+
+void
+user_privilege_add(uint32_t user_id, const char *func_name,
+		   uint32_t func_name_len)
+{
+	/* Grant universe access. */
+	if (is_universe(func_name, func_name_len)) {
+		const struct mh_i32ptr_node_t id_node = {user_id, NULL};
+		struct mh_i32ptr_node_t repl_id_node = {0, NULL};
+		struct mh_i32ptr_node_t *prepl_id_node = &repl_id_node;
+		mh_i32ptr_put(user_runtime_universe, &id_node, &prepl_id_node,
+			      NULL);
+		return;
+	}
+
+	/*  Don't allow to grant built-ins.  */
+	if (tarantool_lua_is_builtin_global(func_name, func_name_len))
+		return;
```

Comment:
It is OK to grant a built-in function. The idea is that granting all the functions (using `'all'` in config or `''` in Lua) shouldn't give rights to execute built-in ones.

Quote from PR #9373:

> If specified, the privilege will be applied only to the given function name, which may be a built-in <...>

Reply:
Now grant_lua_call(<user>, '') allow to call all the global Lua functions except built-in ones.
The only way to call a built-in function is to explicitly grant access to that function.